client/src/components/Visualization/charts/ScatterPlot.js

```diff
@@ -0,0 +1,156 @@
+import { VictoryTheme, 
+         VictoryScatter, 
+         VictoryChart, 
+         VictoryAxis, 
+         VictoryTooltip, 
+         VictoryLabel } from 'victory';
+
+/**
+ * Creates a hexidecimal color string from a string input. 
+ * Taken from here: `https://stackoverflow.com/a/16348977`.
+ * Loosely cleaned up for style and comments.
+ * This *should* return a valid hash, but if it doesn't,
+ * the browser will just display `#000`.
+ * @param {string} str input string 
+ * @return {string}
+ */
+function stringToColor(str) {
+  let hash = 0;
+  let hexColor = '#';
+
+  // Compute a hash from the input string.
+  for (let index = 0; index < str.length; index++) {
+    hash = str.charCodeAt(index) + ((hash << 5) - hash);
+  }
+  
+  // Compute a hex string from the hash.
+  for (let index = 0; index < 3; index++) {
+    let value = (hash >> (index * 8)) & 0xFF;
+    hexColor += ('00' + value.toString(16)).substr(-2);
+  }
+
+  return hexColor;
+}
+
+/**
+ * Turn a firestore snap into a victory formatted
+ * json object that can be charted without further alteration.
+ * @param {object} snap a firestore snap object
+ * @param {string} year year corresponding to the relevant year. 
+ * @return {object}
+ */
+function formatAdvertiserCountSnapshot(snap, year) {
+  const numberOfAds = snap.data().numberOfAds;
+  const advertiser = snap.id;
+
+  const victoryFormattedAdvertiserCount = 
+      {x: parseInt(year), 
+       y: parseInt(numberOfAds), 
+       fill: stringToColor(advertiser),
+       label: `${advertiser} had ${numberOfAds} ads in ${year}.`};
+
+  return victoryFormattedAdvertiserCount;
+}
+
+/**
+ * Returns the exclusive bounds for the range of all data points being charted.
+ * Uses the spread `...` operator.
+ * `https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax` 
+ * @param {List[object]} victoryJsonList a list of objects in a victory-ready 
+ *     format.
+ * @returns {object}
+ */
+function getChartRange(victoryJsonList) {
+  const ranges = victoryJsonList.map(json => json.y);
+
+  return {min: Math.min(...ranges), max: Math.max(...ranges)};
+}
+
+/**
+ * Custom react webhook. Currently calls `dev_aggregates` collection
+ * and queries for the top `year` 
+ * @param {string} year year to query
+ * @param {int} limit maximum number of database documents to return
+ * @return {object}
+ */
+function useAdvertisers(year, queryLimit) {
+  const [advertisers, setAdvertisers] = useState([]);
+
+  useEffect(() => {
+    database.collection("dev_aggregates")
+            .doc(year)
+            .collection("advertisers")
+            .orderBy("numberOfAds", "desc")
+            .limit(queryLimit)
+            .get()
+            .then((snapshots) => {
+      const newAdvertisers = snapshots.docs.map(snap => {
+        return formatAdvertiserCountSnapshot(snap, year);
+      });
+
+      setAdvertisers(newAdvertisers);
+    });
+  }, []);
+
+  return advertisers;
+}
```

Comment:
Have you considered async functions? :)

Reply:
Async in a hook requires you pull out the query into another function, introducing another layer of abstraction etc. Since this isn't even chaining, it's just thening, I think it's ok : )